### DISCUSSION TITLE
ESP32 + 2 CC1101 - send and receive simultaneously

### DISCUSSION THREAD

**Discussion by blok17** [jgromes/RadioLib#335] [Q&A]
Hey @jgromes ,
I am researching with your Library with the usual ESP32 and two CC1101.
I wrote a script to jam and receive at the same time with the two antennas, the configuration of the two CC1101 works perfectly and the jamming antenna too, the problem is the receiving one that outputs a code -6, ERR_RX_TIMEOUT.
Since the script is a bit long, maybe it is better if we quickly chat by email, I got some questions and I don't want to constantly write stuff in the github repo lol.

Also, what protocols does the library support? I am using a radio doorbell (433.92), trying to receive data from the CC1101 with the default script from your examples, but the signal is not detected and I keep receiving a code -6.

Thanks for your time!

**Comment by jgromes** [comment]
Feel free to ask here, that's what Discussions are about.

ERR_RX_TIMEOUT simply means there was nothing received during a call to the blocking `CC1101::receive()`. There's an interrupt-based version which is usually better, see CC1101_Receive_Interrupt.

Regarding the protocols that are supported, there's a couple ham radio protocols/digital modes, such as RTTY or AX.25. It's unlikely the doorbell uses using anything like that, my guess is it will be some simple FSK packet. You will have to record the signal with an SDR and examine it in order to determine the packet format and/or protocol.

**Reply by blok17** [reply]
SO, I implemented the CC1101_Receive_Interrupt script with the usual ESP32+ only one CC1101. I am using your example script plus the following setting: `radio.setRxBandwidth(100);`
If set to 433.92, I am receiving a lot of junk data, like so:
```
Waiting for packet...
CRC error!
Control - Received packet!
[CC1101] Data:		0⸮⸮)⸮⸮⸮⸮UW⸮⸮?$⸮⸮⸮*⸮
CRC error!
CRC error!
Control - Received packet!
[CC1101] Data:		C⸮⸮R`⸮⸮⸮⸮⸮⸮X⸮x⸮⸮(e⸮⸮⸮V⸮T_P	⸮⸮Sao⸮
CRC error!
Control - Received packet!
[CC1101] Data:		j<
CRC error!
Control - Received packet!
[CC1101] Data:		!⸮⸮i`⸮⸮⸮{yT⸮⸮⸮c⸮"7⸮c⸮⸮⸮⸮1⸮
g⸮⸮_ ⸮
Control - Received packet!
[CC1101] Data:		3⸮p⸮⸮⸮f⸮O⸮pг⸮⸮
CRC error!
Control - Received packet!
[CC1101] Data:		$ HS⸮⸮⸮⸮⸮⸮⸮gݧ⸮⸮:W⸮{⸮N⸮ԇ⸮⸮⸮,⸮⸮⸮⸮⸮[:~p⸮⸮֋Qh;x1⸮͖_p⸮⸮X⸮<a⸮̚doD⸮⸮P⸮⸮⸮⸮d⸮⸮o+⸮⸮⸮⸮⸮ۚ!	⸮⸮⸮Q⸮⸮⸮⸮en=⸮⸮0⸮⸮⸮⸮C⸮Nc⸮
h$6A⸮⸮hA?⸮⸮⸮M⸮P⸮⸮⸮'⸮J]C⸮⸮)⸮⸮PG⸮⸮⸮⸮g⸮81⸮⸮⸮⸮⸮⸮⸮⸮6⸮?⸮⸮W⸮ '⸮a⸮⸮ؓ⸮⸮m⸮B⸮⸮ Ƭ8⸮⸮cP⸮Q⸮x⸮⸮E⸮
Control - Received packet!
[CC1101] Data:		⸮⸮⸮⸮⸮⸮⸮⸮⸮⸮⸮#)⸮Q⸮⸮⸮p⸮jDT⸮3JO)
⸮⸮⸮˒[⸮#⸮⸮I⸮⸮⸮⸮,X⸮⸮(⸮r@⸮*߸dw⸮!⸮⸮⸮}n⸮⸮⸮DG⸮⸮⸮2`46f⸮!⸮!⸮y⸮ *⸮k⸮⸮⸮⸮	⸮!)*⸮⸮"⸮S⸮⸮⸮⸮⸮⸮⸮⸮⸮⸮]⸮⸮⸮dm⸮⸮⸮⸮0us⸮⸮M)⸮⸮He⸮⸮⸮!ow⸮⸮⸮(.'⸮ W⸮⸮
CRC error!
Control - Received packet!
[CC1101] Data:		⸮@4A
CRC error!
Control - Received packet!
[CC1101] Data:		⸮⸮a$t⸮k⸮⸮⸮!⸮⸮⸮⸮$OB⸮⸮?K3⸮:⸮⸮?⸮B⸮⸮v?⸮
⸮M⸮5⸮⸮⸮G⸮⸮fA⸮⸮⸮⸮dG⸮⸮z⸮
Control - Received packet!
[CC1101] Data:		 ⸮⸮⸮>K⸮⸮⸮S⸮⸮⸮⸮⸮&⸮Ht֊⸮⸮⸮⸮\>⸮]m/u⸮⸮⸮⸮⸮⸮J⸮0{3⸮⸮⸮f['	⸮`qS+⸮5K⸮⸮⸮5⸮ҰY⸮⸮a⸮⸮⸮⸮8*
⸮w⸮~M⸮{
Control - Received packet!
[CC1101] Data:		⸮0-⸮⸮!⸮!⸮⸮⸮⸮⸮`⸮~B ⸮
⸮⸮⸮y⸮⸮"c⸮n1⸮6⸮#⸮˩5⸮⸮⸮9⸮⸮j⸮Cd⸮
a⸮C=A֧⸮5⸮<E⸮⸮V⸮⸮⸮⸮⸮⸮⸮⸮qT⸮s⸮>P⸮U⸮⸮u⸮ ⸮q⸮⸮⸮⸮91⸮AB⸮I⸮⸮⸮1⸮A⸮Qe⸮⸮⸮⸮V⸮⸮ۑ K⸮b⸮1'⸮⸮⸮Ac⸮ \⸮⸮⸮;⸮&#⸮݅⸮
Control - Received packet!
[CC1101] Data:		⸮8⸮ρ	⸮⸮⸮ө⸮⸮P⸮G⸮⸮⸮⸮I⸮
CRC error!
Control - Received packet!
[CC1101] Data:		u
CRC error!
Control - Received packet!
[CC1101] Data:		⸮⸮jm⸮1g⸮Z⸮A⸮⸮⸮⸮i⸮⸮̑⸮
```
without really being able to detect the packet sent from the doorbell I am testing.

I am monitoring the situation with a HackRF and gqrx in the meanwhile and there is nothing loud at 433.92, everything pretty standard.
It's a pretty random output that comes only if I tune the CC1101 to 433.92 (I tested with 200 MHz and nothing is in output). Surely 433.92 is a busy frequency, but maybe I need to decrease the sensitivity of the receiver? or set something differently?

**Reply by jgromes** [reply]
If you're using the unmodified example, then it's running in packet mode, and it would have to detect - well, packets. That is preamble-sync-payload. It won't detect just noise as a packet. So maybe a loose GDO pin triggering false interrupts is a better explanation. Also, do you have the `setFlag()` interrupt service routine placed in IRAM if you're using ESP32?

Going back to your originial issues - what are you using to power the CC1101? Is it the 3.3V output from ESP32? And if so, is it enough?

**Reply by blok17** [reply]
Here are the tests I run to understand where the problem's coming from:
- I changed pins and jumpers for both the GDO0 pins and nothing changes, same problem.
- I changed the USB power cable - still nothing changed.

I do am implementing the setFlag() function, not sure what you mean when you say "do you have the setFlag() interrupt service routine placed in IRAM if you're using ESP32?".
What I did is I copied the script from `CC1101_Receive_Interrupt ` without really adding anything else, just editing the settings.

About the power issue, don't worry, I will just skip the issue for a sec (I found other ways to make it work) since I'm sure the main problem (junk data outputted from the receiver and receiver not detecting the signal from the doorbell) does not come from there.

**Reply by jgromes** [reply]
On ESP32/ESP8266, all code that is executed within interrupt context must be placed in IRAM. That means marking the function with an attribute:

```c++
void IRAM_ATTR setFlag(void) {
  ...
}
```

>  I'm sure the main problem (junk data outputted from the receiver and receiver not detecting the signal from the doorbell) does not come from there.

I wouldn't be so sure in your case, undervoltage can easily mess up any part of the receiver/demodulator/decoder chain within the CC1101.

Next, are you sure your CC1101s are working? Did you try sending stuff from one to the other (having them connected to separate ESP32s)? If not, you can't be sure they're working at all, much less in this quite unusual setup.

**Reply by blok17** [reply]
So, I added the IRAM_ATTR in the setFlag function (only there, I do not use any other function but loop and begin).

To be sure that the problem does not come from the Power source, I am now powering the esp32 and both the CC1101 with a powerbank this way:
```
Powerbank (5V)--> LM2596 (to 3.3V) --> ESP32 and CC1101
```
and I have the same problem, same random output like so:
```
Control - Received packet!
[CC1101] Data:		A⸮a⸮Wd	
CRC error!
Control - Received packet!
[CC1101] Data:		⸮⸮⸮N#⸮ͮD⸮⸮⸮Gɢ⸮⸮Gԏ⸮qɟ⸮/K⸮R=⸮m⸮A5⸮s⸮⸮@qk⸮Y
CRC error!
Control - Received packet!
[CC1101] Data:		⸮;qu[⸮⸮1lϴWfT
Control - Received packet!
[CC1101] Data:		I4w⸮s~֫%⸮⸮2dE⸮@⸮Ig⸮z}ˏ⸮⸮)2⸮⸮_⸮Ic⸮%~⸮=⸮K;zB⸮⸮⸮	O⸮JT1⸮ib⸮ )⸮H
Control - Received packet!
[CC1101] Data:		⸮N⸮T⸮K⸮⸮hy⸮⸮\⸮⸮OJ⸮<⸮Y⸮U⸮[⸮A⸮⸮ ⸮⸮L
CRC error!
Control - Received packet!
[CC1101] Data:		~̰$(GD⸮"⸮⸮⸮)⸮⸮⸮⸮⸮քKA`L⸮⸮⸮b⸮⸮⸮⸮w"⸮R`⸮
CRC error!
Control - Received packet!
[CC1101] Data:		32⸮⸮I^)Ơ⸮B⸮.Q@⸮⸮
CRC error!
Control - Received packet!
[CC1101] Data:		QȒ>dt⸮⸮⸮H}⸮]B`?⸮@&⸮⸮⸮⸮
CRC error!
CRC error!
```
No idea where it's getting those random packets.

I am sure that the CC1101 are working since I transmitted from both of them, but I have not been able to receive anything but junk data (as above).

I'll try to detach everything and just link the esp32 to a cc1101, loading the simplest firmware I found to just receive a signal. Let's see where I get. 
Keep you posted.

EDIT: here's the output with the simplest program to receive and only one cc1101 attached (note that for this one I changed every single jumper with a new one and connected everything directly, no breadboard used):
```
[CC1101] Received packet!
[CC1101] Data:		u⸮\⸮'⸮⸮⸮d⸮P⸮⸮⸮g8⸮=⸮⸮I⸮⸮q⸮⸮=+`⸮\9Y⸮⸮w>T⸮⸮⸮⸮⸮+⸮⸮g!⸮#6⸮⸮L~yk⸮⸮⸮@8_.⸮⸮⸮B⸮⸮i⸮⸮Y⸮^E⸮,j⸮⸮ޢ⸮⸮⸮^'⸮
⸮⸮⸮I
[CC1101] RSSI:		-51.50 dBm
[CC1101] LQI:		96
CRC error!
[CC1101] Received packet!
[CC1101] Data:		҅⸮m.H,⸮⸮⸮~a$⸮{⸮⸮⸮l_⸮⸮ܧje⸮
⸮⸮⸮efXG9⸮⸮⸮
[CC1101] RSSI:		-95.50 dBm
[CC1101] LQI:		127
[CC1101] Received packet!
[CC1101] Data:		⸮߆⸮e⸮Ĉjyosix%8⸮⸮$Ɣ⸮A⸮⸮⸮(g⸮<Ls}K90⸮棫-⸮\&⸮⸮c"⸮ Jh⸮(⸮"⸮8⸮⸮⸮C⸮S⸮
~o⸮⸮⸮⸮I⸮4⸮)즈⸮⸮⸮}B⸮|P'⸮Y⸮Ӓ폷⸮⸮l⸮b⸮_
[CC1101] RSSI:		-95.50 dBm
[CC1101] LQI:		127
CRC error!
[CC1101] Received packet!
[CC1101] Data:		_⸮⸮ak⸮0⸮YF:⸮⸮2⸮\;⸮1⸮X-⸮⸮⸮⸮p<⸮⸮b{4[⸮G⸮⸮⸮⸮⸮CB⸮⸮&]⸮G⸮⸮⸮⸮Fq⸮+⸮⸮`⸮⸮⸮⸮⸮Wai⸮⸮⸮}w+QV⸮⸮S[N8W⸮ݵ"⸮⸮9⸮⸮⸮c
⸮(VQ⸮+⸮⸮
[CC1101] RSSI:		-92.00 dBm
[CC1101] LQI:		127
[CC1101] Received packet!
[CC1101] Data:		\⸮$⸮>I5i⸮⸮⸮4R⸮⸮n⸮M=⸮jh⸮⸮/nx_⸮⸮i*⸮B⸮mn⸮⸮⸮d⸮B⸮%d~⸮⸮D@⸮⸮@N⸮W敯b⸮z⸮⸮փǸ⸮⸮⸮'⸮W7⸮⸮⸮Ќ⸮])f
Ł	⸮=⸮⸮J'⸮⸮⸮In⸮x⸮X⸮⸮⸮0⸮&⸮ղ⸮⸮⸮ػy*⸮⸮⸮⸮a⸮=⸮/⸮⸮S'⸮v⸮Xl⸮⸮9/q⸮9⸮G⸮{⸮⸮⸮⸮⸮_⸮⸮!⸮⸮⸮⸮eCk?"⸮/⸮C⸮⸮⸮
[CC1101] RSSI:		-92.00 dBm
[CC1101] LQI:		127

```

At this point I am also thinking if maybe the connections of the CC1101 are right.
I am using `VCC,SI,SO,CSN,GND,SCK,GDO0,GDO2`.

Really dunno what/where to debug now..

**Reply by blok17** [reply]
I'll open an issue related to this discussion since seems ESP32 + CC1101 + RadioLib have some issues receiving the signal.
Thanks for the help for now!

**Reply by jgromes** [reply]
I got curious about this issue myself, and made a simple setup: Arduino Uno with CC1101 as the transmitter, and NodeMCU ESP32S with CC1101 as the receiver. I did indeed see some garbage packets, but not a lot of them, and they could always be easily distinguished by very high LQI and very low RSSI:

Valid packets: LQI < 8, RSSI > -50 dBm
Garbage packets: LQI >> 80, RSSI < -100 dBm

There might be some configuration issue that's making the receiver too sensitive, but that's most likely platform independent, so unrelated to ESP32 being used. Also I don't really have the time to search for this.

**Reply by blok17** [reply]
I see. What do you suggest to have it working? should I change board? and if so, which one do you suggest it's better to use two CC1101 at the same time?

**Reply by jgromes** [reply]
I'm not all that familiar with your project, was there a specific reason you decided to use CC1101? If not, SX1278 might be a better alternative.

**Reply by blok17** [reply]
Not really, I just need to send and receive from one device at the same time, the device and the antennas do not matter.. 
I see that the SX1278 manage only frequencies from 137.0 MHz to 525.0 MHz and I need to go a bit higher too..


Unfortunately, the only two boards I have here are the ESP32 and the Raspberry Pi Pico (not supported yet :/) so I will get an ESP8266s to double check if the problem is coming from the board or not.

**Comment by blok17** [comment]
So I kept digging, I found out that the working/not working of the antennas is quite random. Here's what's going on.
Code I run to set up the two CC1101:
```

...

CC1101 jammer = new Module(27, 17, RADIOLIB_NC, 26, SPI, SPISettings(2000000, MSBFIRST, SPI_MODE0));   
CC1101 control = new Module(13, 12, RADIOLIB_NC, 14, SPI, SPISettings(2000000, MSBFIRST, SPI_MODE0));

...


void begin() {
  int state1;
  int state2;
  while(1) {
    Serial.print("Configuring Jammer: ");
    SPI.begin();
    state1 = jammer.begin(frequency1, 32.0, 60.0, 250.0, 10, 32);
    if(state1 == ERR_NONE){
        Serial.print("DONE");Serial.println("");
        break;
    }
    else{
        Serial.print(state1);Serial.println(""); 
    }
  }
  while(state2 == ERR_NONE) {
    Serial.println("Configuring Control: ");
    SPI.begin();
    state2 = control.begin(frequency2, 32.0, 60.0, 250.0, 10, 32);
    if(state1 == ERR_NONE){
      Serial.println("DONE");Serial.println("");
      break;
    }
  }
  
}
```
I call the `begin()` in the `setup()`.
When I run it, if I run it for the first time, it works immediately. If I reset the ESP32 (with the button), it starts outputting error -16 or -2 (randomly), sometimes it stops and works after a while, other times it just keeps going with the -16/-2 without stopping.
I notice that every time I change the GDO0 pin rewiring, it works the first time, no errors. If i start resetting, same thing described above happens (-16/-2).

Any idea why or how to stabilise the thing?

EDIT: Also I am noticing that if I let it run while it gives errors and I wait some minutes before resetting through the button, it then works. I am exploding here, help me lol

**Reply by jgromes** [reply]
Couple of thoughts:

1. Try enabling debug mode, it should give you some more insight into the -16 error codes (which is an SPI write fail).
2. You're continually initializing SPI in a loop, and I have no idea if ESP32 SPI driver is smart enough to handle this. The usual use-case is to initialize once on startup and then let it be.
3. Both error codes can also be related to hardware, which would explain the randomness. I just spent several weekends debugging an issue with Si4432, which turned out to be largely caused by a faulty breadboard, so I suggest thoroughly checking connections.

**Reply by blok17** [reply]
**Other update**: while I have the -2/-16 output going on, if I detach the 3V3 pin (power going from the ESP32 to the two antennas) from the board and reattach it back, errors stop and the setup finishes successfully.

(I'll keep the thread updated so if anyone else has the same problem he/she can follow my path.)

**Reply by jgromes** [reply]
What are you using as the power source? The 3.3V regulator on your ESP32 probably can't output enough current for the ESP32 as well as two CC1101 radios.

**Reply by blok17** [reply]
Yes, I disconnected the screen I was using and I think I solved the problem (I am a dumbass, I am powering the thing using a USB cable through a usb hub where I have my keyboard connected, prolly not enough power to handle all of that!) 

I'll keep going, the antennas are now working quite well, thanks for the help for now, but I do know that I will find some other issue in the path I might need you for.

Thanks Jan, appreciate your work here.
